### PR TITLE
Add tensor table in ModelDef and use it for jit script serialization and deserialization

### PR DESCRIPTION
**Pull request by houseroad**
As we discussed, the tensors in the torch script will be associated with the tensor data in the serialized file. So let's add a table of tensor (actually it's a repeated TensorProto filed) in the ModelDef. TensorProto.name will be the id.

#accept2ship

### CODE REVIEW

**Review by dzhulgakov — APPROVED** [review]

File: torch/csrc/jit/export.cpp [hunk 1]
```diff
@@ -465,32 +465,146 @@ class MethodEncoder : public EncoderBase {
...
 
+void MethodEncoder::EncodeMethod(
+    std::string* torch_script,
+    const script::Method& method,
+    const std::string prefix) {
+  onnx::ModelProto model_proto;
+  model_proto.set_doc_string("THIS PROTO IS NOT STANDARD ONNX");
+  auto* node_proto = model_proto.mutable_graph()->add_node();
+  node_proto->set_name(prefix + method.name());
+  if (method.is_optimized()) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: is it stacked on previous diff?

File: torch/csrc/jit/export.cpp [hunk 2]
```diff
@@ -465,32 +465,146 @@ class MethodEncoder : public EncoderBase {
...
+    // mark that this method was optimized
+    node_proto->set_domain("optimized");
+  }
+
+  // We store the schema string in the docstring.
+  node_proto->set_doc_string(getExportableSchemaStringForMethod(method));
+
+  // Store member_inputs of Method in input
+  for (auto &member_input : method.params()) {
+    auto it = serializer_->paramMap().find(member_input);
```
Comment (dzhulgakov) [inline torch/csrc/jit/export.cpp]: nit: it's a nicer API if (de)serializer just exposes "lookupParam" instead of exposing the entire map.

**Review by dzhulgakov — COMMENTED** [review]
Looks good!

Do we have any unittests that cover the in-place Tensors?

File: torch/csrc/jit/export.cpp (hunk 1 above)
Comment (houseroad) [reply torch/csrc/jit/export.cpp]: nope, will rebase one on the other.

File: torch/csrc/jit/export.cpp (hunk 2 above)
Comment (houseroad) [reply torch/csrc/jit/export.cpp]: agree, will do it

**Comment by houseroad** [comment]
We have tests which cover the tensor type attributes :-)

**Pull request closed without merging**